File: backend/utils/role_required.py
from functools import wraps
from flask_jwt_extended import get_jwt
from flask import jsonify


def role_required(role):
    """Decorator — requires JWT and checks `role` claim."""
    def decorator(fn):
        @wraps(fn)
        def wrapper(*args, **kwargs):
            claims = get_jwt()
            if not claims or claims.get('role') != role:
                return jsonify({'msg': f'forbidden - requires {role} role'}), 403
            return fn(*args, **kwargs)

        return wrapper

    return decorator

File: backend/utils/roleRequired.js
const jwt = require('jsonwebtoken');

const verifyJWT = (req, res, next) => {
  const authHeader = req.headers.authorization;

  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    return res.status(401).json({ msg: 'missing or invalid authorization header' });
  }

  const token = authHeader.substring(7);

  try {
    const decoded = jwt.verify(token, process.env.JWT_SECRET_KEY || 'super-secret-key');
    req.user = decoded;
    next();
  } catch (err) {
    return res.status(401).json({ msg: 'invalid or expired token' });
  }
};

const roleRequired = (requiredRole) => {
  return (req, res, next) => {
    if (req.user.role !== requiredRole) {
      return res.status(403).json({ msg: `forbidden - requires ${requiredRole} role` });
    }
    next();
  };
};

module.exports = { verifyJWT, roleRequired };
